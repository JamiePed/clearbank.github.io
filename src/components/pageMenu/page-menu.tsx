import React, { useState, useRef, useLayoutEffect, useMemo, useEffect } from 'react'
import throttle from 'lodash.throttle'
import flatten from 'lodash.flatten'
import kebabCase from 'lodash.kebabcase'

import { hasWindow } from 'src/utils/browser.services'

import * as Types from './page-menu.types'
import * as Styles from './page-menu.styles'

const mapHeadingToComponent = (heading: any) => {
  const currentHash = hasWindow() ? window.location.hash : ''
  const href = `#${heading.id}`
  const isActive = currentHash === href

  switch(heading.tag) {
    case 'H1':
      return (
        <Styles.FirstLevelLink href={href} isActive={isActive}>
          {heading.title}
        </Styles.FirstLevelLink>
      )
    case 'H2':
      return (
        <Styles.BaseLink href={href} isActive={isActive}>
          {heading.title}
        </Styles.BaseLink>
    )
    case 'H3':
      return (
        <Styles.ThirdLevelLink href={href} isActive={isActive}>
          {heading.title}
        </Styles.ThirdLevelLink>
      )
    case 'H4':
      return (
        <Styles.FourthLevelLink href={href} isActive={isActive}>
          {heading.title}
        </Styles.FourthLevelLink>
      )
    default:
      return null;
  }
}


const PageMenu: React.FC<Types.PageMenuProps> = () => {
  const [headings, setHeadings] = useState([]);

  useEffect(() => {
    if (!hasWindow()) {
      return
    }
console.log({ headings })
  const rawHeadings = Array.from(document.querySelectorAll('.page .page-menu-entry'))
    .map(mdxTitle => ({
      tag: mdxTitle.tagName,
      parent: mdxTitle.closest('.page').id,
      title: mdxTitle.textContent,
      id: mdxTitle.id
    }))
    .filter((entry: any) => entry.id !== '')

    setHeadings(rawHeadings)
  }, [])

  return (
    <Styles.Container id='pageMenu'>
      <Styles.Title>On this page:</Styles.Title>
      <Styles.List>
        {headings.map(heading => (
          <li key={heading.id}>
            {mapHeadingToComponent(heading)}
          </li>
        ))}
      </Styles.List>
    </Styles.Container>
  )
}

export default PageMenu
