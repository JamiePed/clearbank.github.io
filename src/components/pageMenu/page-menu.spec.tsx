import React from 'react'
import { render } from '@testing-library/react'
import '@testing-library/jest-dom'
import 'jest-styled-components'

import Component from './page-menu'

import * as Types from './page-menu.types'

let component: any = null
const defaultProps: Types.PageMenuProps = {
  data: {}
}

jest.mock('src/utils/browser.services', () => ({
  hasWindow: () => {
    return true
  }
}))

// mock window.machMedia
Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: jest.fn().mockImplementation(query => ({
    matches: false,
    media: query,
    onchange: null,
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn()
  }))
})

beforeEach(() => {
  component = render(<Component {...defaultProps} />)
})

describe('Component', () => {
  test('matches snapshot', () => {
    expect(component.asFragment()).toMatchSnapshot()
  })
})
