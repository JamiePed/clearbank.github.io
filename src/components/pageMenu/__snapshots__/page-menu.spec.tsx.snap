// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component matches snapshot 1`] = `
<DocumentFragment>
  .c0 {
  position: -webkit-sticky;
  position: sticky;
  top: 220px;
  right: 0;
  max-height: calc(100vh - 220px);
  overflow-y: scroll;
  color: #fff;
  height: 100%;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  padding: 50px 23px 0px 23px;
  border-left: 3px solid #747474;
}

.c0:scrollbar {
  display: none;
}

.c1 {
  margin: 0;
  color: #000000;
  font-weight: 600;
  font-size: 32px;
  cursor: default;
  line-height: 38px;
}

.c2 {
  list-style: none;
  margin-top: 15px;
}

<div
    class="c0"
    id="pageMenu"
  >
    <h5
      class="c1"
    >
      On this page:
    </h5>
    <ul
      class="c2"
    />
  </div>
</DocumentFragment>
`;
